nflow-engine/src/main/java/io/nflow/engine/internal/dao/ExecutorDao.java

```diff
@@ -164,27 +163,44 @@ private void updateWithPreparedStatement(String sql) {
   }
 
   public List<WorkflowExecutor> getExecutors() {
-    return jdbc.query("select * from nflow_executor where executor_group = ? order by id asc", new RowMapper<WorkflowExecutor>() {
-      @Override
-      public WorkflowExecutor mapRow(ResultSet rs, int rowNum) throws SQLException {
-        int id = rs.getInt("id");
-        String host = rs.getString("host");
-        int pid = rs.getInt("pid");
-        DateTime started = sqlVariants.getDateTime(rs, "started");
-        DateTime active = sqlVariants.getDateTime(rs, "active");
-        DateTime expires = sqlVariants.getDateTime(rs, "expires");
-        DateTime stopped = sqlVariants.getDateTime(rs, "stopped");
-        return new WorkflowExecutor(id, host, pid, executorGroup, started, active, expires, stopped);
-      }
-    }, executorGroup);
+    return jdbc.query("select id, host, pid, started, active, expires, stopped, recovered from nflow_executor where "
+        + getExecutorGroupCondition() + " order by id asc", (rs, rowNum) -> {
+          int id = rs.getInt("id");
+          String host = rs.getString("host");
+          int pid = rs.getInt("pid");
+          var started = sqlVariants.getDateTime(rs, "started");
+          var active = sqlVariants.getDateTime(rs, "active");
+          var expires = sqlVariants.getDateTime(rs, "expires");
+          var stopped = sqlVariants.getDateTime(rs, "stopped");
+          var recovered = sqlVariants.getDateTime(rs, "recovered");
+          return new WorkflowExecutor(id, host, pid, executorGroup, started, active, expires, stopped, recovered);
+        });
   }
 
   public void markShutdown() {
```

Comment:
Isn't this supposed to be graceful shutdown that does not need any recovery? At least the old code did not recover any workflows from executors closed with this method.

Reply:
Yes, unless the graceful shutdown times out. Then we need to recover.